### DISCUSSION TITLE
Backward compatibility without tears

### DISCUSSION THREAD

**Discussion by Totktonada** [tarantool/tarantool#6912] [Ideas]
The usual way to handle compatibility problems is to introduce an option and leave the old behaviour by default. It is not always the perfect way.

Sometimes we want to keep the old behaviour for existing applications and offer the new one by default for new applications. For example the old behaviour is known to be problematic / less safe / does not correspond to expectations of a user. A user not always read all the documentation and often assume good defaults.

I had an idea: introduce a 'compatibility profile' thing. Basically it is a global options table:

```yaml
tarantool> tarantool = require('tarantool')
tarantool> tarantool.compat
---
- json_escape_forward_slash:
    current: true
    brief: Whether to escape the forward slash symbol '/' using a backslash in a json.encode()
      result. The old and the new behaviour produce a result, which is compatible
      with the JSON specification. However most of other JSON encoders don't escape
      the forward slash, so we consider the new behaviour as more safe one.
    doc: https://github.com/tarantool/tarantool/wiki/compat_json_escape_forward_slash
    old: true
    new: false
...
```

With helper functions, getters and setters:

```yaml
tarantool> tarantool.compat.candidates()
---
- json_escape_forward_slash:
    current: true
    brief: <...>
    doc: https://github.com/tarantool/tarantool/wiki/compat_json_escape_forward_slash
    old: true
    new: false
...
tarantool> tarantool.compat.json_escape_forward_slash = false
---
...
tarantool> tarantool.compat.candidates()
---
- []
...
tarantool> tarantool.compat.dump()
---
- json_escape_forward_slash: false
...
tarantool> tarantool.compat.restore({json_escape_forward_slash = false})
---
...
```

(The table format, function names and so on are just to illustrate the idea and likely not the best choice.)

Just global table is not the all story. To make it a solution for the problem, we should implement the following:

* The table with brief help, old and new values, a link to more in-depth explanation.
* Getters and setters for the values.
* The functions to dump and restore a profile.
* A function to show what options can be updated to the new behaviour.
* A log warning at start for old values. Possibly also CLI / UI warnings.
* An entrypoint example with the `tarantool.compat.restore()` call in the documentation. The same for application examples / templates (including ones provided by a CLI).

----

Doubts.

An application may stitch many very different parts: modules, frameworks, integrations with external services, domain specific logic. It may be not easy (and even impossible in practice) to verify whether all parts of the application are ready for enabling a new behaviour in a built-in module.

OTOH, the idea has no sense if we'll set different compatibility options for different parts of the applications: this way it is the same as just a runtime option and does not introduce any simplification for a user. In other words, this way we lack of one place to tweak the compatibility options.

It is not a simple problem and I doubt that it has an always working / automatic solution. However there is a light beam. Since we have a documentation URL for each compatibility entry, we can:

* Update the documentation with all known cases, when the option should not be changed: say if a particular version range of a particular module is used.
* We can add more steps to the 'update pathway' section even if we had no such steps in the mind initially: as result of an investigation of a user's problem.

----

What is also worth to say there. There is an opinion that all compatibility problems were resolved by [semver](https://semver.org). It is far beyond the truth for library alike projects: ones, which are base ground for other libraries and applications.

A project maintainers have the responsibility to do their best to don't let existing users update its code (or validate, whether it needs an update) too often. Depending of a nature of the project, this period varies from half of a year to several years.

Semantic versioning provides a way to *deliver* information that an action is required to update the software, nothing more. It does not provide an indulgence to change anything you want to change (if you have users). The responsibility can't disappear, because some magic numbers are changed.

In practice, even a new major version can't change anything. Well, it breaks a particular set of functionality to provide something better. But not more: otherwise it is pure nightmare for a user, to understand whether (s)he can update or can't.

----

Inspired by https://github.com/tarantool/tarantool/issues/6200.

**Comment by Lord-KA** [comment]
# Using compatibility module

Compat is run automatically at the start of a tarantool session. You can access it as ``tarantool.compat``. It lists all compatibility options controlled by compat:

    tarantool> compat = require('tarantool').compat
    ---
    ...
        
    tarantool> compat
    ---
    - json_escape_forward_slash: new
    - option_2: old
    - new_option_disable_vinyl: default (old/new)
    ...


* To change options you can do it directly:
        
        tarantool> compat.json_escape_forward_slash = 'old'
        ---
        ...

   or pass a list of such options

        tarantool> compat{json_escape_forward_slash = 'new', option_2 = 'default'}
        ---
        ...

   all options can be set as 'old'/'new'/'default'. 'default' marks the option as unselected. 
* You can check if a specific option has been set correctly:

       tarantool> compat.new_option_disable_vinyl
       ---
       - current: 'old'
         default: 'new'
         brief: <...>
       ...

* You can set a specific option to default:

        tarantool> compat.restore({'option_2'})
        ---
        ...

        tarantool> compat.option_2
        ---
        - current: 'old'
          default: 'old'
          brief: <...>
        ...

   or restore default for every option:

        tarantool> compat.reset()
        ---
        ...

* To list all unselected options, run:

       tarantool> compat.candidates()
       ---
       - option_2:
           brief: <...>
           current: default
           default: old
         json_escape_forward_slash:
           brief: <...>
           current: default
           default: old
         new_option_disable_vinyl:
           brief: <...>
           current: default
           default: new
       ...


* To get a command that can set up compat with the same options as current, run:

        tarantool> compat{json_escape_forward_slash = 'new', option_2 = 'new'}
        ---
        ...

        tarantool> compat.dump()
        ---
        - |+
          require('tarantool').compat({
              json_escape_forward_slash = 'new',
              option_2                  = 'new',
              new_option_disable_vinyl  = 'default',
          })
        
        ...
        
    Or you can get a setup with all options set as 'new'/'old'/'default' by passing the state as an argument:
        
        tarantool> compat.dump('new')
        ---
        - |+
          require('tarantool').compat({
              json_escape_forward_slash = 'new',
              option_2                  = 'new',
              new_option_disable_vinyl  = 'new',
          })
        
        ...

* You can also add options during runtime with `add_option()` that requires list of tables with the following:
    - name (string)
    - default ('new'/'old')
    - brief (explanation of the option, can be multiline string)
    - obsolete (true/false -  is the option frozen on 'new')
    - action function (arg - boolean as if behavior is 'new' == true or 'old' == false, changes the behavior accordingly)
             
        ```
        tarantool> compat.add_option{
                         name = 'option_4',
                         default = 'new',
                         brief = "<...>",
                         obsolete = false,                       -- You can explicitly mark option as non-obsolete.
                         action = function(val)
                              print(("option_4 action was called with val = %d!"):format(val)) 
                         end
                   }
        option_4 postaction was called with val = 1!
        ```

   Note that action of a new option is called with its defaults right after it is added (before the next one in the list).

* **Hot reload:** you can change an existing option in runtime using `add_option()`, it will update all the fields but keep currently selected behavior if any and run `action` with it.

* To add option to compat src directly, add table with option description (same as for `add_option()`) and function to action table. Every built-in option must have a Tarantool wiki page with detailed description, list of compatibility problems and some tips on detecting problems in codebase. For example, see `json_escape_forward_slash` patch (gh-6200).

* Make sure to set up compat if needed at the very beginning of a program or module, right after the require calls.

**Comment by Lord-KA** [comment]
### Summary

The usual way to handle compatibility problems is to introduce an option for new behavior and leave the old one by default. It is not always the perfect way.

Sometimes we want to keep the old behavior for existing applications and offer the new one by default for new ones. For example, the old behavior is known to be problematic / less safe / does not correspond to user expectations. In contrast, the user doesn’t always read all the documentation and often assumes good defaults. It was decided to introduce a compatibility module to provide a direct way to deprecate unwanted behavior.

 

`tarantool.compat` module is basically a global options table with additional verbose interface and helper functions. There are three stages of changing behavior:

- old behavior by default
- new behavior by default
- new behavior is frozen and the old behavior is removed

on the first two stages, user can toggle options via the interface and change the behavior according to his needs, on the last stage old behavior is removed from codebase and option is marked as obsolete. As `compat` is a global instance, options can be hardcoded into it or added in runtime e.g. by external module.

Options will be switched to the next stage in major releases, this way developers would be able to adapt to the new standard behavior and test it before switching to the next release. If something is broken by the new tarantool version, developer would still have a way to fix it by a simple config change (explicitly select old behavior). 

**Consider example below:**

- Option `json_esc_slash` is introduced in `2.12` minor release. Default is set to `'old'` but developer can utilize new behavior or test updated behavior by switching it manually to `'new'`.
- In `3.0` — next major release, `json_esc_slash`'s default is switched to `'new'`. Now developers that didn’t manage to adapt to new behavior will be able to switch option to `'old'` and fix their module in the future.
- In `4.0` `json_esc_slash` is marked as `obsolete` and old behavior is no longer accessible. Developers are forced to use new behavior.


### Tutorial

The options list is serialized in the interactive console with additional details for user convenience:

- All non-obsolete options in order `new` → `old` → `default` .
- Serialization returns array-like table with tables `{<option> = <value>}` 
- The result of `compat` serialization can still be indexed as a normal key-value table.

```yaml
tarantool> compat = require('tarantool').compat
---
...
    
tarantool> compat
---
- - json_escape_forward_slash: new
- - option_2: old
- - option_default_old: default (old)
- - option_default_new: default (new)
...
```

1. **Listing options details:**
- `current` — the state of the option
- `default` — the default state of the option
- `brief` — text options description with a link to [tarantool.wiki](https://github.com/tarantool/tarantool/wiki/compat) in a [specific format](https://github.com/tarantool/tarantool/wiki/compat%3Aoptio_name).

```yaml
tarantool> compat.option_default_new
 ---
 - current: old
   default: new
   brief: <...>
   obsolete: false      -- explicit
 ...
```

1. **Changing option value:**
- You can do it directly, or by passing a table with option-value.
- Assignable values are `'new'` , `'old'` and `'default'`.

```yaml
tarantool> compat.json_escape_forward_slash = 'old'
  ---
  ...

tarantool> compat{json_escape_forward_slash = 'new', option_2 = 'default'}
  ---
  ... 
```

1. **Restoring defaults:**
- By setting `'default'` value to it

```yaml
tarantool> compat.option_2 = 'default'
  ---
  ...
tarantool> compat.option_2
---
- current: default
- default: new
- brief: <...>
...
```

1. **Getting tarantool.compat setup with `compat.dump()`:**

```yaml
tarantool> compat({
         >     obsolete_set_explicitly = 'new',
         >     option_set_old = 'old',
         >     option_set_new = 'new'
         > })
---
...

tarantool> compat
---
- -	option_set_old: old
- - option_set_new: new
- - option_default_old: default (old)
- - option_default_new: default (new)
- - obsolete_option_default: default (new)
- - obsolete_set_explicitly: new
...

# nil does output obsolete unset options as 'default'
tarantool> compat.dump()
---
- require('tarantool').compat({
			option_set_old          = 'old',
			option_set_new          = 'new',
			option_default_old      = 'default',
			option_default_new      = 'default',
			obsolete_option_default = 'default', -- obsolete since X.Y.Z
      obsolete_set_explicitly = 'new',     -- obsolete since X.Y.Z
	})
...
# 'current' is same as nil with default set to current values
tarantool> compat.dump('current')
---
- require('tarantool').compat({
			option_set_old          = 'old',
			option_set_new          = 'new',
			option_default_old      = 'old',
			option_default_new      = 'new',
			obsolete_option_default = 'new',     -- obsolete since X.Y.Z
      obsolete_set_explicitly = 'new',     -- obsolete since X.Y.Z
	})
...
# 'new' outputs obsolete as 'new'.
tarantool> compat.dump('new')
---
- require('tarantool').compat({
			option_set_old          = 'new',
			option_set_new          = 'new',
			option_default_old      = 'new',
			option_default_new      = 'new',
			obsolete_option_default = 'new',     -- obsolete since X.Y.Z
      obsolete_set_explicitly = 'new',     -- obsolete since X.Y.Z
	})
...
# 'old' outputs obsolete options as 'new'.
tarantool> compat.dump('old')
---
- require('tarantool').compat({
			option_set_old          = 'old',
			option_set_new          = 'old',
			option_default_old      = 'old',
			option_default_new      = 'old',
			obsolete_option_default = 'new',     -- obsolete since X.Y.Z
      obsolete_set_explicitly = 'new',     -- obsolete since X.Y.Z
	})
...
# 'default' does output obsolete options as default.
tarantool> dump('default')
---
- require('tarantool').compat({
			option_set_old          = 'default',
			option_set_new          = 'default',
			option_default_old      = 'default',
			option_default_new      = 'default',
			obsolete_option_default = 'default', -- obsoleted since X.Y.Z
      obsolete_set_explicitly = 'default', -- obsoleted since X.Y.Z
	})
...
```

1. **Setting all options to a specific value with `compat.dump()`:**
- use `compat.dump()` to get a specific configuration
- copy and paste it into console (or use `loadstring()`)

```yaml
tarantool> compat.dump('new')
---
- require('tarantool').compat({
      option_2 = 'new',
      json_escape_forward_slash = 'new',
        })
...
tarantool> require('tarantool').compat({
      option_2 = 'new',
      json_escape_forward_slash = 'new',
        })
---
...

tarantool> compat
---
- - json_escape_forward_slash: new
- - option_2: new
...
```

1. **Adding an option during runtime:**
- User must provide a table with:
    - name (string)
    - default (`’new’` / `’old’`)
    - brief (explanation of the option, can be multiline string)
    - obsolete (`’X.Y.Z’` / `false` / `nil` -  tarantool version that marked option as obsolete. When it isn’t provided, option is treated as non-obsolete)
    - action function (argument - boolean as if behavior is 'new' == true or 'old' == false, changes the behavior accordingly)
    - run_action (`true` / `false` / `nil`) if `add_options` should run action afterwards, `false` by default
- Option hot reload:
    - you can change an existing option in runtime using `add_option()`, it will update all the fields but keep currently selected behavior if any. The new `action` will be called afterwards.

```yaml
tarantool> compat.add_option{
                 name = 'option_4',
                 default = 'new',
                 brief = "<...>",
                 obsolete = false,        -- You can explicitly mark option as non-obsolete.
                 action = function(is_new)
                      print(("option_4 action was called with is_new = %s!"):format(is_new)) 
                 end,
                 run_action = true
           }
option_4 postaction was called with is_new = true!
---
...

tarantool> compat.add_option{             -- hot reload of option_4
                 name = 'option_4',
                 default = 'old',         -- different default
                 brief = "<...>",
                 action = function(is_new)
                      print(("new option_4 action was called with is_new = %s!"):format(is_new)) 
                 end
           }
---
...         -- action is not called by default
```

**Comment by Totktonada** [comment]
https://github.com/tarantool/tarantool/pull/7060 is pushed into 2.11 and will come with first 2.11 release.
